File: c/src/data/list/list.c
/*
 * MIT License
 *
 * Copyright(c) 2020 Pedro Henrique Penna <pedrohenriquepenna@gmail.com>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#include <assert.h>
#include <stdlib.h>
#include <utils.h>
#include "list.h"

/**
 * @brief Linked list.
 */
struct list
{
	int x;              /**< Element   */
	struct list *nextp; /**< Next Node */
};

/**
 * The list_create() function creates a linked list. A dummy head node
 * for the new list is created and returned.
 */
struct list *list_create(void)
{
	struct list *headp;

	headp = smalloc(sizeof(struct list));

	/* Initialize list. */
	headp->nextp = headp;

	return (headp);
}

/**
 * The list_destroy() function destroys the linked list pointed to by @p
 * listp.
 */
int list_destroy(struct list *listp)
{
	/* Sanity check. */
	assert(listp != NULL);
	assert(listp == listp->nextp);

	/* House keeping. */
	free(listp);

	return (0);
}

/**
 * The list_insert() function inserts the element @p x in the list
 * pointed to by @p listp. A new list node is created to store the
 * element, and this node is inserted after the the node pointed to
 * by @p listp.
 */
int list_insert(struct list *listp, int x)
{
	struct list *nodep;

	/* Sanity check. */
	assert(listp != NULL);

	/* Creates a new list node. */
	nodep = smalloc(sizeof(struct list));

	/* Initialize node. */
	nodep->x = x;

	/* Dequeue node. */
	nodep->nextp = listp->nextp;
	listp->nextp = nodep;

	return (0);
}

/**
 * The list_remove() function removes the next element in the list
 * pointed to by @p listp. The remove element is returned.
 */
int list_remove(struct list *listp)
{
	int x;
	struct list *nodep;

	/* Sanity check. */
	assert(listp != NULL);
	assert(listp->nextp != listp);

	/* Dequeue node. */
	nodep = listp->nextp;
	listp->nextp = nodep->nextp;

	x = nodep->x;

	/* House keeping. */
	free(nodep);

	return (x);
}

/**
 * @brief The list_next() function returns the next node in the list
 * pointed to by @p listp.
 */
struct list *list_next(struct list *listp)
{
	/* Sanity check */
	assert(listp != NULL);
	assert(listp->nextp != NULL);

	return (listp->nextp);
}

File: go/src/data/list/list.go
/*
 * Copyright(c) 2020 Pedro Henrique Penna <pedrohenriquepenna@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package list

import (
	"fmt"
)

// List node.
type listNode struct {
	key  int
	next *listNode
}

// List
type list struct {
	length int
	head   listNode
	tail   *listNode
}

// Creates a linked list.
func New() *list {
	l := new(list)

	l.length = 0
	l.head.next = nil
	l.tail = &l.head

	return l
}

// Deletes a linked list.
func (l *list) Delete() {

	// Release all nodes.
	for w := l.head.next; w != nil; {
		p := w
		w = w.next
		p.next = nil
	}

	l.length = 0

	// Release memory
	l.head.next = nil
	l.tail = nil
}

// Returns the length of a linked list.
func (l *list) Length() int {
	return l.length
}

// Returns the head node of a list.
func (l *list) Head() *listNode {
	return &l.head
}

// Returns the tail node of a list.
func (l *list) Tail() *listNode {
	return l.tail
}

// Returns the next node in a list.
func (node *listNode) Next() *listNode {
	return node.next
}

// Inserts a new node to a linked list (after an existing node).
func (l *list) InsertAfter(node, newNode *listNode) {
	newNode.next = node.next
	node.next = newNode
	l.length++
}

// Inserts a new node to a linked list (at the front).
func (l *list) PushFront(node *listNode) {

	node.next = l.head.next
	l.head.next = node

	// Update tail.
	if l.tail == &l.head {
		l.tail = node
	}

	l.length++
}

// Removes a node from a linked list (at the back).
func (l *list) PushBack(node *listNode) {

	node.next = nil
	l.tail.next = node
	l.tail = node

	// Update head.
	if l.head.next == nil {
		l.head.next = node
	}

	l.length++
}

// Removes the next node in a linked list.
func (l *list) RemoveNext(node *listNode) {

	// Update tail
	if node.next == l.tail {
		l.tail = node
	}

	node.next = node.next.next
	l.length--
}

// Prints a list.
func (l *list) Print() {
	for w := l.head.next; w != nil; w = w.next {
		fmt.Print(w.key, "->")
	}
	fmt.Println("nil")
}

// Tests the linked list implementation.
func Test() {
	l := New()

	// Insert some nodes in the list.
	for i := 0; i < 10; i++ {
		node := new(listNode)
		node.key = i
		l.PushFront(node)
	}

	l.Print()

	// Remove some nodes from the list.
	for w := l.Head(); w.Next() != nil; w = w.Next() {
		l.RemoveNext(w)
	}

	l.Print()

	// Insert some nodes in the list.
	for i := 1; i < 10; i += 2 {
		node := new(listNode)
		node.key = i
		l.PushBack(node)
	}

	l.Print()

	// Remove some nodes from the list.
	for w := l.Head(); w.Next() != nil; w = w.Next() {
		l.RemoveNext(w)
	}

	l.Print()

	for i := 0; i < 10; i += 2 {
		node := new(listNode)
		node.key = 20
		l.InsertAfter(l.Head(), node)
	}

	l.Print()
}
